File: pcces-go/internal/storage/sqlite/conversion_long_jobs.go
package sqlite

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	errx "github.com/liuxb99/pcces-web/pcces-go/internal/platform/errors"
)

type ConversionLongJob struct {
	ID string `json:"id"`
	JobType string `json:"job_type"`
	Status string `json:"status"`
	Progress int `json:"progress"`
	Stage string `json:"stage"`
	Payload map[string]any `json:"payload"`
	Result map[string]any `json:"result,omitempty"`
	Error map[string]any `json:"error,omitempty"`
	CancelRequested bool `json:"cancel_requested"`
	CreatedBy string `json:"created_by"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
	RowVersion int64 `json:"row_version"`
}

type ConversionLongJobRepository struct{ store *Store }
func NewConversionLongJobRepository(store *Store) *ConversionLongJobRepository { return &ConversionLongJobRepository{store:store} }

func (r *ConversionLongJobRepository) Create(ctx context.Context,id,jobType,actor string,payload map[string]any)(ConversionLongJob,error){
	jobType=strings.ToUpper(strings.TrimSpace(jobType));if jobType!="IMPORT"&&jobType!="EXPORT"{return ConversionLongJob{},errx.New(errx.CodeInvalidArgument,"job_type must be IMPORT or EXPORT","P4-G-JOB")}
	b,_:=json.Marshal(payload);now:=time.Now().UTC().Format(time.RFC3339Nano)
	_,err:=r.store.db.ExecContext(ctx,`INSERT INTO conversion_long_jobs(id,job_type,status,progress,stage,payload_json,cancel_requested,created_by,created_at,updated_at,row_version) VALUES(?,?,?,?,?,?,?,?,?,?,1)`,id,jobType,"QUEUED",0,"QUEUED",string(b),0,actor,now,now)
	if err!=nil{return ConversionLongJob{},err};return r.Get(ctx,id)
}
func (r *ConversionLongJobRepository) Get(ctx context.Context,id string)(ConversionLongJob,error){
	var v ConversionLongJob;var payload,result,errorJSON sql.NullString;var cancel int
	err:=r.store.db.QueryRowContext(ctx,`SELECT id,job_type,status,progress,stage,payload_json,result_json,error_json,cancel_requested,created_by,created_at,updated_at,row_version FROM conversion_long_jobs WHERE id=?`,id).Scan(&v.ID,&v.JobType,&v.Status,&v.Progress,&v.Stage,&payload,&result,&errorJSON,&cancel,&v.CreatedBy,&v.CreatedAt,&v.UpdatedAt,&v.RowVersion)
	if err==sql.ErrNoRows{return v,errx.New(errx.CodeNotFound,"conversion job not found","P4-G-JOB")};if err!=nil{return v,err};v.CancelRequested=cancel!=0;_ = json.Unmarshal([]byte(payload.String),&v.Payload);if result.Valid{_ = json.Unmarshal([]byte(result.String),&v.Result)};if errorJSON.Valid{_ = json.Unmarshal([]byte(errorJSON.String),&v.Error)};return v,nil
}
func (r *ConversionLongJobRepository) Advance(ctx context.Context,id string,rowVersion int64,progress int,stage,status string,result,errorValue map[string]any)(ConversionLongJob,error){
	current,err:=r.Get(ctx,id);if err!=nil{return current,err};if current.Status=="COMPLETED"||current.Status=="FAILED"||current.Status=="CANCELLED"{return current,errx.New(errx.CodeConflict,"terminal job cannot advance","P4-G-JOB")};if rowVersion!=current.RowVersion{return current,errx.New(errx.CodeConflict,"stale conversion job row_version","P4-G-JOB")};if progress<current.Progress||progress>100{return current,errx.New(errx.CodeInvalidArgument,"progress must be monotonic and between 0 and 100","P4-G-JOB")};status=strings.ToUpper(strings.TrimSpace(status));if status==""{if progress==100{status="COMPLETED"}else{status="RUNNING"}};if current.CancelRequested{status="CANCELLED";progress=current.Progress};if status!="RUNNING"&&status!="COMPLETED"&&status!="FAILED"&&status!="CANCELLED"{return current,errx.New(errx.CodeInvalidArgument,"invalid job status","P4-G-JOB")};rb,_:=json.Marshal(result);eb,_:=json.Marshal(errorValue);now:=time.Now().UTC().Format(time.RFC3339Nano);res,err:=r.store.db.ExecContext(ctx,`UPDATE conversion_long_jobs SET status=?,progress=?,stage=?,result_json=?,error_json=?,updated_at=?,row_version=row_version+1 WHERE id=? AND row_version=?`,status,progress,defaultString(stage,status),nullableJSON(rb,result),nullableJSON(eb,errorValue),now,id,rowVersion);if err!=nil{return current,err};n,_:=res.RowsAffected();if n!=1{return current,errx.New(errx.CodeConflict,"conversion job update conflict","P4-G-JOB")};return r.Get(ctx,id)
}
func (r *ConversionLongJobRepository) Cancel(ctx context.Context,id string,rowVersion int64)(ConversionLongJob,error){
	current,err:=r.Get(ctx,id);if err!=nil{return current,err};if current.Status=="COMPLETED"||current.Status=="FAILED"||current.Status=="CANCELLED"{return current,nil};if rowVersion!=current.RowVersion{return current,errx.New(errx.CodeConflict,"stale conversion job row_version","P4-G-JOB")};now:=time.Now().UTC().Format(time.RFC3339Nano);res,err:=r.store.db.ExecContext(ctx,`UPDATE conversion_long_jobs SET status='CANCELLED',stage='CANCELLED',cancel_requested=1,result_json=NULL,error_json=NULL,updated_at=?,row_version=row_version+1 WHERE id=? AND row_version=?`,now,id,rowVersion);if err!=nil{return current,err};n,_:=res.RowsAffected();if n!=1{return current,fmt.Errorf("cancel conflict")};return r.Get(ctx,id)
}
func nullableJSON(b []byte,v map[string]any) any { if v==nil{return nil};return string(b) }

File: api/conversion_long_jobs.py
"""Phase 4 long-running conversion jobs with progress and cancellation."""
from __future__ import annotations

import json
from datetime import datetime, timezone
from uuid import uuid4

from flask import Blueprint, jsonify, request
from sqlalchemy import Boolean, Column, DateTime, Integer, MetaData, String, Table, Text, and_, select

metadata = MetaData()
conversion_long_jobs = Table(
    "conversion_long_jobs", metadata,
    Column("id", String(100), primary_key=True),
    Column("job_type", String(20), nullable=False),
    Column("status", String(20), nullable=False),
    Column("progress", Integer, nullable=False),
    Column("stage", String(100), nullable=False),
    Column("payload_json", Text, nullable=False),
    Column("result_json", Text, nullable=True),
    Column("error_json", Text, nullable=True),
    Column("cancel_requested", Boolean, nullable=False, default=False),
    Column("created_by", String(100), nullable=False),
    Column("created_at", DateTime(timezone=True), nullable=False),
    Column("updated_at", DateTime(timezone=True), nullable=False),
    Column("row_version", Integer, nullable=False, default=1),
)

TERMINAL = {"COMPLETED", "FAILED", "CANCELLED"}


class ConversionLongJobService:
    def __init__(self, engine):
        self.engine = engine
        metadata.create_all(engine)

    def create(self, body: dict, actor: str) -> dict:
        job_type = str(body.get("job_type", "")).strip().upper()
        if job_type not in {"IMPORT", "EXPORT"}:
            raise ValueError("job_type must be IMPORT or EXPORT")
        now = datetime.now(timezone.utc)
        job_id = str(uuid4())
        with self.engine.begin() as conn:
            conn.execute(conversion_long_jobs.insert().values(
                id=job_id, job_type=job_type, status="QUEUED", progress=0, stage="QUEUED",
                payload_json=json.dumps(body.get("payload") or {}, ensure_ascii=False, sort_keys=True),
                result_json=None, error_json=None, cancel_requested=False,
                created_by=actor, created_at=now, updated_at=now, row_version=1,
            ))
        return self.get(job_id)

    def get(self, job_id: str) -> dict:
        with self.engine.connect() as conn:
            row = conn.execute(select(conversion_long_jobs).where(conversion_long_jobs.c.id == job_id)).mappings().first()
        if not row:
            raise LookupError("conversion job not found")
        item = dict(row)
        for key in ("payload_json", "result_json", "error_json"):
            value = item.pop(key)
            item[key.removesuffix("_json")] = json.loads(value) if value else None
        item["created_at"] = item["created_at"].isoformat()
        item["updated_at"] = item["updated_at"].isoformat()
        item["deep_link"] = f"/app/conversions/jobs/{job_id}"
        return item

    def advance(self, job_id: str, body: dict) -> dict:
        current = self.get(job_id)
        if current["status"] in TERMINAL:
            raise ValueError("terminal job cannot advance")
        requested = int(body.get("row_version", 0))
        if requested != int(current["row_version"]):
            raise RuntimeError("stale conversion job row_version")
        progress = int(body.get("progress", current["progress"]))
        if progress < int(current["progress"]) or progress > 100:
            raise ValueError("progress must be monotonic and between 0 and 100")
        now = datetime.now(timezone.utc)
        status = str(body.get("status") or ("COMPLETED" if progress == 100 else "RUNNING")).upper()
        if current["cancel_requested"]:
            status, progress = "CANCELLED", int(current["progress"])
        if status not in {"RUNNING", "COMPLETED", "FAILED", "CANCELLED"}:
            raise ValueError("invalid job status")
        result = body.get("result") if status == "COMPLETED" else None
        error = body.get("error") if status == "FAILED" else None
        with self.engine.begin() as conn:
            changed = conn.execute(conversion_long_jobs.update().where(and_(
                conversion_long_jobs.c.id == job_id,
                conversion_long_jobs.c.row_version == requested,
            )).values(
                status=status, progress=progress, stage=str(body.get("stage") or status),
                result_json=json.dumps(result, ensure_ascii=False, sort_keys=True) if result is not None else None,
                error_json=json.dumps(error, ensure_ascii=False, sort_keys=True) if error is not None else None,
                updated_at=now, row_version=requested + 1,
            ))
            if changed.rowcount != 1:
                raise RuntimeError("conversion job update conflict")
        return self.get(job_id)

    def cancel(self, job_id: str, row_version: int) -> dict:
        current = self.get(job_id)
        if current["status"] in TERMINAL:
            return current
        if int(row_version) != int(current["row_version"]):
            raise RuntimeError("stale conversion job row_version")
        now = datetime.now(timezone.utc)
        with self.engine.begin() as conn:
            changed = conn.execute(conversion_long_jobs.update().where(and_(
                conversion_long_jobs.c.id == job_id,
                conversion_long_jobs.c.row_version == row_version,
            )).values(
                status="CANCELLED", stage="CANCELLED", cancel_requested=True,
                result_json=None, error_json=None, updated_at=now, row_version=row_version + 1,
            ))
            if changed.rowcount != 1:
                raise RuntimeError("conversion job cancel conflict")
        return self.get(job_id)


def build_conversion_long_job_blueprint(service: ConversionLongJobService, resolve_user_id):
    bp = Blueprint("conversion_long_jobs", __name__, url_prefix="/api/conversions/jobs")

    @bp.post("")
    def create_job():
        actor = resolve_user_id()
        if actor is None:
            return jsonify({"code": "UNAUTHORIZED"}), 401
        try:
            return jsonify(service.create(request.get_json(silent=True) or {}, str(actor))), 201
        except ValueError as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400

    @bp.get("/<job_id>")
    def get_job(job_id: str):
        try:
            return jsonify(service.get(job_id))
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404

    @bp.post("/<job_id>/advance")
    def advance(job_id: str):
        try:
            return jsonify(service.advance(job_id, request.get_json(silent=True) or {}))
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404
        except RuntimeError as exc:
            return jsonify({"code": "CONFLICT", "detail": str(exc)}), 409
        except ValueError as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400

    @bp.post("/<job_id>/cancel")
    def cancel(job_id: str):
        try:
            body = request.get_json(silent=True) or {}
            return jsonify(service.cancel(job_id, int(body.get("row_version", 0))))
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404
        except RuntimeError as exc:
            return jsonify({"code": "CONFLICT", "detail": str(exc)}), 409

    return bp
